File: nefelis/lingen_gf2.py
"""
Computation of linear generator for 1 or several sequences over GF(2)

This is the analogue for lingen for arithmetic modulo 2.
Currently it only implements the quadratic (original) version,
which does not require polynomial multiplication.

In this implementation matrices of GF(2) polynomials are represented
by matrices of big integers, using Python ints.

The implementation follows the conventions of:

Emmanuel Thomé
Fast computation of linear generators for matrix sequences and application to the block Wiedemann algorithm.
ISSAC '01: Proceedings of the 2001 international symposium on Symbolic and algebraic computation, Jul 2001, London, Ontario, Canada. pp.323-331,
https://inria.hal.science/inria-00517999
"""

import logging
import random
import time

import flint
import numpy
import numpy.typing as npt

from nefelis.backends.kompute.lingen_gf2 import mslgdc as mslgdc_gpu

DEBUG_LINGEN = False

DEBUG_CHECK_LINGEN = False

DEBUG_TIMINGS = True

logger = logging.getLogger("lingen")


def lingen(sequences, N: int):
    """
    Compute a (simultaneous) linear generating polynomial for input sequences.

    The polynomial sum a[i] X^i is such that:
       sum a[i] seq[k+i] = 0 for all k

    N: the expected degree of polynomial (matrix dimension in Block Wiedemann)
    """
    m = len(sequences)
    if m == 1:
        seq = sequences[0]
        assert len(seq) > 2 * N + 1
        ctx = flint.fmpz_mod_poly_ctx(2)
        pol = ctx.minpoly(seq)
        return flint.nmod_poly([int(b) for b in pol], 2)

    assert all(len(s) > N + N // m + 16 for s in sequences)

    M = numpy.zeros((m, 1), dtype=object)
    for i, seq in enumerate(sequences):
        M[i, 0] = sum(1 << i for i, b in enumerate(seq) if b)

    for t0 in range(m, m + 4):
        for _ in range(10):
            F = numpy.zeros((1, 1 + m), dtype=object)
            seen = set([0])
            for i in range(m):
                r = random.getrandbits(t0)
                while r in seen:
                    r = random.getrandbits(t0)
                seen.add(r)
                F[0, i] = r
            F[0, m] = 1 << t0

            # Compute matrix product E = (M @ F) >> m
            E = numpy.zeros((m, 1 + m), dtype=object)
            for i in range(m):
                for j in range(1 + m):
                    E[i, j] = clmul(M[i, 0], F[0, j]) >> t0

            E0 = (E[:, :m] & 1).astype(numpy.uint8)
            js = set()
            E0det = 1
            for i in range(m):
                try:
                    j0 = next(j for j in range(m) if j not in js and E0[i, j])
                except StopIteration:
                    E0det = 0
                    break
                js.add(j0)
                for j in range(m):
                    if j not in js and E0[i, j]:
                        E0[:, j] ^= E0[:, j0]
            if E0det != 0:
                break
        if E0det != 0:
            break

    assert E0det != 0
    # print("Selected t0 =", t0)
    # Constraint (Thomé, section 2.3.2)
    # Eij[0] must be a matrix of rank m

    delta = [t0 for _ in range(m + 1)]
    EXTRA_ITERS = 16

    if DEBUG_LINGEN:
        for i, j in numpy.ndindex(*E.shape):
            print(f"before E[{i},{j}]", hex(E[i, j]))

    P = mslgdc(E, delta, N + N // m + EXTRA_ITERS)
    if DEBUG_LINGEN:
        for i, j in numpy.ndindex(*E.shape):
            print(f"after E[{i},{j}]", hex(E[i, j]))

        print("P")
        print(P)
        print(delta)

    # We just want FP[i, 0]
    fp0 = 0
    for j in range(1 + m):
        fp0 ^= clmul(P[j, 0], F[0, j])

    fp0bits = bin(fp0)[2:]
    return [int(b) for b in fp0bits]


def lingen_mat(mats, N: int):
    """
    Compute a linear generating (matrix) polynomial for a matrix of sequences.

    The polynomial sum a[i] X^i is such that:
       sum a[i] seq[k+i] = 0 for all k

    N: the expected degree of polynomial (matrix dimension in Block Wiedemann)
    """
    m = len(mats)
    n = len(mats[0])

    assert all(len(s) > N // m + N // n + 16 for row in mats for s in row)

    M = numpy.zeros((m, n), dtype=object)
    tm0 = time.monotonic()
    for i in range(m):
        for j in range(n):
            seq = mats[i][j]
            M[i, j] = sum(1 << i for i, b in enumerate(seq) if b)
    dt0 = time.monotonic() - tm0

    tm1 = time.monotonic()
    for t0 in range(m, m + 4):
        for _ in range(10):
            # To obtain a nonsingular matrix, choose a larger number of columns
            EXTRA = 16
            F = numpy.zeros((n, m + EXTRA), dtype=object)
            for i in range(n):
                for j in range(m + EXTRA):
                    r = random.getrandbits(t0)
                    F[i, j] = r

            # Compute matrix product E = (M @ F) >> m
            E = numpy.zeros((m, m + EXTRA), dtype=object)
            for i in range(m):
                for j in range(m + EXTRA):
                    eij = 0
                    for k in range(n):
                        eij ^= clmul(M[i, k], F[k, j]) >> t0
                    E[i, j] = eij

            # Now find a nonsingular submatrix
            E0 = (E & 1).astype(numpy.uint8)
            js = set()
            sing = False
            for i in range(m):
                try:
                    j0 = next(j for j in range(m + EXTRA) if j not in js and E0[i, j])
                except StopIteration:
                    sing = True
                    break
                js.add(j0)
                for j in range(m):
                    if j not in js and E0[i, j]:
                        E0[:, j] ^= E0[:, j0]

            # Extract submatrix and augment with X^t0 Identity
            Esub = numpy.zeros((m, n + m), dtype=object)
            Fsub = numpy.zeros((n, n + m), dtype=object)
            for i, j in enumerate(sorted(js)):
                Esub[:, i] = E[:, j]
                Fsub[:, i] = F[:, j]
            for i in range(n):
                Fsub[i, i + m] = 1 << t0
                Esub[:, i + m] = M[:, i]
            E, F = Esub, Fsub

            if not sing:
                break
        if not sing:
            break

    dt1 = time.monotonic() - tm1

    assert not sing
    # print("Selected t0 =", t0)
    # Constraint (Thomé, section 2.3.2)
    # Eij[0] must be a matrix of rank m

    delta = [t0 for _ in range(m + n)]
    EXTRA_ITERS = 16

    if DEBUG_LINGEN:
        for i, j in numpy.ndindex(*E.shape):
            print(f"before E[{i},{j}]", hex(E[i, j]))

    tm2 = time.monotonic()
    P = mslgdc_gpu(E, delta, N // m + N // n + EXTRA_ITERS)
    dt2 = time.monotonic() - tm2
    if DEBUG_LINGEN:
        for i, j in numpy.ndindex(*E.shape):
            print(f"after E[{i},{j}]", hex(E[i, j]))

        print("P")
        print(P)
        print(delta)

    tm3 = time.monotonic()
    # Compute (F @ P)
    FP = numpy.zeros((n, n + m), dtype=object)
    degree = 0
    for i in range(m):
        for j in range(n + m):
            fij = 0
            for k in range(n + m):
                fij ^= clmul(P[k, j], F[i, k])
            if j < n:
                degree = max(degree, fij.bit_length())
            FP[i, j] = fij
    dt3 = time.monotonic() - tm3

    if DEBUG_CHECK_LINGEN:
        b = N // m + N // n + 1
        # The nxn square matrix D (left half of FP) must be such
        # that MD is small.
        for i in range(m):
            for j in range(n):
                mij = 0
                for k in range(n):
                    mij ^= clmul(M[i, k], FP[k, j])
                mij_trunc = mij & ((1 << b) - 1)
                if max(i, j) <= 2:
                    print(f"D[{i},{j}] length {FP[i, j].bit_length()}")
                    print(f"MD[{i},{j}] length {mij_trunc.bit_length()}")
                assert mij_trunc.bit_length() < N // m + 64, mij_trunc.bit_length()

    tm4 = time.monotonic()
    poly = [numpy.zeros((m, n), dtype=numpy.uint8) for _ in range(degree)]
    for i in range(m):
        for j in range(n):
            fpij = FP[i, j]
            assert fpij.bit_length() <= degree
            for k in range(degree):
                poly[k][i, j] = (fpij >> k) & 1
    dt4 = time.monotonic() - tm4

    if DEBUG_TIMINGS:
        logger.debug(
            f"lingen: import {dt0:.3f}s, init {dt1:.3f}s, mslgdc {dt2:.3f}s, fp {dt3:.3f}s, fini {dt4:.3f}s"
        )
    return poly


def clmul(x: int, y: int, karatsuba=10000, block=True) -> int:
    """
    Polynomial multiplication over GF(2). y is usually smaller than x

    >>> clmul(57, 73)
    4017
    >>> clmul(123456789, 1234)
    135009063098
    """
    nx, ny = x.bit_length(), y.bit_length()
    # Basic Karatsuba for large inputs
    if karatsuba and min(nx, ny) > karatsuba:
        M = (nx + ny) // 4
        mask = (1 << M) - 1
        xlo, ylo = x & mask, y & mask
        xhi, yhi = x >> M, y >> M
        A = clmul(xlo, ylo, karatsuba, block)
        B = clmul(xlo ^ xhi, ylo ^ yhi, karatsuba, block)
        C = clmul(xhi, yhi, karatsuba, block)
        return A ^ ((A ^ B ^ C) << M) ^ (C << (2 * M))
    if nx < ny:
        x, y = y, x
        nx, ny = ny, nx
    if block and ny > 64:
        # Block multiplication
        mk = [0, x]
        for i in (1, 2, 3):
            xi = x << i
            mk += [m ^ xi for m in mk]
        z = 0
        for k in range(0, ny, 4):
            z ^= mk[(y >> k) & 15] << k
    else:
        z = 0
        for k in range(ny):
            if y & 1:
                z ^= x << k
            y >>= 1
    return z


def mslgdc(E, delta, b):
    P = None
    timings = None
    if DEBUG_TIMINGS:
        timings = [0.0, 0.0, 0.0]
    for _ in range(b):
        P, delta = lingen_step(E, delta, P=P, timings=timings)
        # print(delta)
    if DEBUG_TIMINGS:
        logger.debug(
            f"mslgdc: sort {timings[0]:.3f}s elim {timings[1]:.3f}s shift {timings[2]:.3f}s"
        )
    return P


def lingen_step(
    E: npt.NDArray, delta: list[int], P=None, timings=None
) -> tuple[npt.NDArray, list[int]]:
    # Advance by 1 degree using Gauss elimination (ALGO1 in [Thomé])
    # E is modified in place and divided by X
    # P is optionally modified in place
    m, mn = E.shape
    if P is None:
        P = numpy.array(
            [[(1 if i == j else 0) for j in range(mn)] for i in range(mn)],
            dtype=object,
        )
    # Sort columns
    t0 = time.monotonic()
    for i in range(mn):
        argmin = min(range(i, mn), key=lambda idx: (delta[idx], idx))
        if i < argmin:
            delta[i], delta[argmin] = delta[argmin], delta[i]
            for j in range(mn):
                P[j, i], P[j, argmin] = P[j, argmin], P[j, i]
            for j in range(m):
                E[j, i], E[j, argmin] = E[j, argmin], E[j, i]
    if timings:
        timings[0] += time.monotonic() - t0

    # Eliminate without increasing degree (col[j] is cancelled using col[j0<j])
    t1 = time.monotonic()
    nonzero = set()
    for i in range(m):
        # Find first nonzero element of row i
        try:
            j0 = next(j for j in range(mn) if (E[i, j] & 1) and j not in nonzero)
        except StopIteration:
            continue
        nonzero.add(j0)
        assert E[i, j0] & 1 == 1
        # Use it to eliminate next columns
        for j in range(j0 + 1, mn):
            if E[i, j] & 1:
                for h in range(m):
                    E[h, j] ^= E[h, j0]
                for h in range(mn):
                    P[h, j] ^= P[h, j0]
    if timings:
        timings[1] += time.monotonic() - t1
    if DEBUG_LINGEN:
        for i, j in numpy.ndindex(*E.shape):
            if E[i, j] & 1:
                assert j in nonzero
    # Now each row has 1 nonzero coefficient
    t2 = time.monotonic()
    for j in nonzero:
        delta[j] += 1
        for i in range(mn):
            P[i, j] <<= 1
    # Combine multiplication by X and division by X
    for j in range(mn):
        if j not in nonzero:
            for i in range(m):
                assert E[i, j] & 1 == 0
                E[i, j] >>= 1
    if timings:
        timings[2] += time.monotonic() - t2
    # debug("P", P)
    return P, delta


def benchmark_m1():
    import time
    import sys

    sys.set_int_max_str_digits(10000)

    for N in (1000, 2000, 4000, 20000):
        random.seed(42)
        taps = [random.getrandbits(1) for _ in range(min(200, N))]

        # A single sequence of size 2N+O(1) for reference
        seqlong = [random.getrandbits(1) for _ in range(N)]
        for _ in range(N + 64):
            t = sum(a * seqlong[-N + i] for i, a in enumerate(taps)) & 1
            seqlong.append(t)

        t0 = time.monotonic()
        refpoly = lingen([seqlong], N)
        dt = time.monotonic() - t0
        print(f"lingen for {N=} m=1 in {dt:.3f}s")
        assert refpoly.degree() <= N, len(refpoly)

        want = (N * [0]) + [1]
        for i in range(len(taps)):
            want[i] = taps[i]
        assert flint.nmod_poly(want, 2) % refpoly == 0

        for k in range(200, N - 200, 37):
            assert sum(refpoly[i] * seqlong[k + i] for i in range(N + 1)) == 0

        for m in (2, 3, 4, 16, 32, 64):
            # m sequences of size N+N/m+O(1) for testing
            seqs = []
            for _ in range(m):
                seq = [random.getrandbits(1) for _ in range(N)]
                for _ in range(N // m + 80):
                    t = sum(a * seq[-N + i] for i, a in enumerate(taps)) & 1
                    seq.append(t)
                seqs.append(seq)

            t0 = time.monotonic()
            minpoly = lingen(seqs, N)
            dt = time.monotonic() - t0
            print(f"lingen for {N=} {m=} n=1 in {dt:.3f}s")
            print(f"=> polynomial has degree {len(minpoly) - 1}")
            # print(minpoly)

            pol2 = flint.nmod_poly(minpoly, 2)
            # assert pol2 % refpoly == 0


def benchmark_mn():
    import time

    for N in (1000, 2000, 4000, 20000):
        # random.seed(42)
        taps = [random.getrandbits(1) for _ in range(min(200, N))]

        # for m in (2, 3, 4, 16):
        for m in (4, 16, 32, 64):
            # mn sequences of size N/m+N/n+O(1) for testing
            # First we build m sequences
            seqs = []
            for _ in range(m):
                seq = [random.getrandbits(1) for _ in range(N)]
                for _ in range(N):
                    t = sum(a * seq[-N + i] for i, a in enumerate(taps)) & 1
                    seq.append(t)
                seqs.append(seq)
            # Then we compute m projections
            projs = []
            for _ in range(m):
                idx = random.sample(list(range(N // 2)), 20)
                projs.append(idx)

            mats = [[] for _ in range(m)]
            for i in range(m):
                seq = seqs[i]
                for j, idx in enumerate(projs):
                    seqij = [
                        sum(seq[k + j] for j in idx) & 1
                        for k in range(N // m + N // m + 64)
                    ]
                    mats[j].append(seqij)

            t0 = time.monotonic()
            _minpoly = lingen_mat(mats, N)
            dt = time.monotonic() - t0
            print(f"lingen for {N=} {m=} n={m} in {dt:.3f}s")


if __name__ == "__main__":
    logging.basicConfig(level=logging.DEBUG)
    logging.getLogger().setLevel(level=logging.DEBUG)
    benchmark_mn()

File: rust/src/lingen_gf2.rs
#![allow(non_snake_case)]

//! Computation of linear generators for GF(2) matrix sequences
//! The implementation follows the conventions of:
//!
//! Emmanuel Thomé
//! Fast computation of linear generators for matrix sequences and application to the block Wiedemann algorithm.
//! ISSAC '01: Proceedings of the 2001 international symposium on Symbolic and algebraic computation, Jul 2001, London, Ontario, Canada. pp.323-331,
//! https://inria.hal.science/inria-00517999
//!
//! All matrices are represented in column-major order.
//! Only square sizes m=n=32 is supported in pratice.

use num_bigint::BigUint;
use rayon::iter::IntoParallelIterator;
use rayon::prelude::ParallelIterator;

use crate::gf2x;

/// Compute a transition matrix such that (E x P)[0] = 0
/// and delta is a valid profile.
///
/// E has shape (m,m+n) represented by m integers with (m+n) bits
fn lingen_step32<const M: usize, const MN: usize>(
    E: &mut [[u32; M]; MN],
    delta: &mut [u32; MN],
    P: &mut [[u32; MN]; MN],
    iter: usize,
) {
    // Sort columns
    for i in 0..MN {
        let ii = if iter > 0 {
            // delta is almost sorted (deltaprev + {0,1} where deltaprev is sorted)
            let mut ii = i;
            for j in (i + 1)..MN {
                if delta[j] < delta[i] {
                    ii = j;
                    break;
                }
                if delta[j] > delta[i] {
                    break; // delta[jj > j] is always >= delta[i]
                }
            }
            debug_assert_eq!(ii, (i..MN).min_by_key(|&idx| delta[idx as usize]).unwrap());
            ii
        } else {
            (i..MN).min_by_key(|&idx| delta[idx as usize]).unwrap()
        };
        if i < ii {
            // Swap columns
            delta.swap(i, ii);
            P.swap(i, ii);
            E.swap(i, ii);
        }
    }
    debug_assert!((1..MN).all(|i| delta[i - 1] <= delta[i]));
    // Elimination
    let mut nonzero = [false; MN];
    for i in 0..M {
        let Some(j0) = (0..MN)
            .filter(|&j| E[j as usize][i] & 1 == 1 && !nonzero[j as usize])
            .next()
        else {
            continue;
        };
        nonzero[j0] = true;
        debug_assert!(E[j0][i] & 1 == 1);
        for j in (j0 + 1)..MN {
            if E[j][i] & 1 == 1 {
                for h in 0..M {
                    E[j][h] ^= E[j0][h];
                }
                for h in 0..MN {
                    P[j][h] ^= P[j0][h];
                }
            }
            debug_assert!(E[j][i] & 1 == 0);
        }
    }
    for j in 0..MN {
        if nonzero[j] {
            debug_assert!((0..M).any(|i| E[j][i] & 1 == 1));
            delta[j] += 1;
            for i in 0..MN {
                P[j][i] <<= 1;
            }
        } else {
            debug_assert!((0..M).all(|i| E[j][i] & 1 == 0));
            for i in 0..M {
                E[j][i] >>= 1;
            }
        }
    }
}

const MSLGDC_THRESHOLD: usize = 32;

/// Divide and conquer algorithm for MSLGDC.
pub(crate) fn mslgdc<const M: usize, const MN: usize>(
    E: &mut [[BigUint; M]; MN],
    delta: &[u32; MN],
    b: usize,
    iter: usize,
) -> Box<[[BigUint; MN]; MN]> {
    if b < MSLGDC_THRESHOLD {
        // Iterative variant
        let mut P = [[0; MN]; MN];
        for i in 0..MN {
            P[i][i] = 1;
        }
        let mut esmall: [[u32; M]; MN] =
            std::array::from_fn(|j| std::array::from_fn(|i| (&E[j][i]).try_into().unwrap()));
        let delta_before = delta.iter().copied().sum::<u32>();
        let mut delta = *delta;
        for i in 0..b {
            lingen_step32(&mut esmall, &mut delta, &mut P, i);
        }
        let delta_after = delta.iter().copied().sum::<u32>();
        assert!(delta_after - delta_before <= (M * b) as u32);
        for i in 0..M {
            for j in 0..MN {
                E[j][i] = esmall[j][i].into();
            }
        }
        return Box::new(P.map(|col| col.map(BigUint::from)));
    }
    if iter > 0 {
        // Invariant: E has degree < b
        debug_assert!(
            E[..]
                .iter()
                .all(|col| col.iter().all(|eij| eij.bits() <= b as u64)),
            "b={b} iter={iter}"
        );
    }
    // WARNING: large arrays must not be put on stack (e.g. if M=64 and MN=128)

    // Low degree size
    let mask = (BigUint::ONE << (b / 2)) - BigUint::ONE;
    let maskhi = (BigUint::ONE << b) - BigUint::ONE;
    let mut EL: Box<[[BigUint; M]; MN]> = Box::new(std::array::from_fn(|j| {
        std::array::from_fn(|i| &E[j][i] & &mask)
    }));
    let PL = mslgdc::<M, MN>(&mut EL, delta, b / 2, iter);

    // Compute (E @ PL)[b-b/2:]
    // We compute columns in parallel
    // EL is not needed anymore
    let mut ER = EL;
    let cols: Vec<[BigUint; M]> = (0..MN)
        .into_par_iter()
        .map(|j| {
            std::array::from_fn(|i| {
                let mut colij = BigUint::ZERO;
                for k in 0..MN {
                    // Eik * Pkj
                    let mij = gf2x::mul(&E[k][i], &PL[j][k]);
                    // Truncate to bits b/2..b
                    colij ^= (mij & &maskhi) >> (b / 2);
                }
                colij
            })
        })
        .collect();
    for (j, colj) in cols.into_iter().enumerate() {
        ER[j] = colj;
    }

    // Delta = Delta * PL
    let delta_r: [u32; MN] = std::array::from_fn(|j| {
        (0..MN)
            .filter(|&i| PL[j][i] != BigUint::ZERO)
            .map(|i| delta[i] + PL[j][i].bits() as u32 - 1)
            .max()
            .unwrap()
    });
    //println!("{} {:?}", iter + b / 2, delta_r);

    let PR = mslgdc::<M, MN>(&mut ER, &delta_r, b - b / 2, iter + b / 2);
    drop(ER);

    // Now compute P = PL @ PR
    // We compute columns in parallel
    let mut P: Box<[[BigUint; MN]; MN]> = Box::new(std::array::from_fn(|_| {
        std::array::from_fn(|_| BigUint::ZERO)
    }));
    let cols: Vec<[BigUint; MN]> = (0..MN)
        .into_par_iter()
        .map(|j| {
            std::array::from_fn(|i| {
                let mut colij = BigUint::ZERO;
                for k in 0..MN {
                    // PLik * PRkj
                    let mij = gf2x::mul(&PL[k][i], &PR[j][k]);
                    colij ^= mij;
                }
                // P has degree <= b
                debug_assert!(colij.bits() <= b as u64 + 1);
                colij
            })
        })
        .collect();
    for (j, colj) in cols.into_iter().enumerate() {
        P[j] = colj;
    }
    P
}
